chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java

```diff
@@ -1424,47 +918,63 @@ public void rebuildActorBvh() {
     refresh();
   }
 
-  private int calculateOctreeOrigin(Collection<ChunkPosition> chunksToLoad, boolean centerOctree) {
+  /**
+   * Find the bounds in blocks for a collection of chunks.
+   */
+  static BlockBounds calculateBounds(Collection<ChunkPosition> chunks) {
     int xmin = Integer.MAX_VALUE;
     int xmax = Integer.MIN_VALUE;
     int zmin = Integer.MAX_VALUE;
     int zmax = Integer.MIN_VALUE;
-    for (ChunkPosition cp : chunksToLoad) {
-      if (cp.x < xmin) {
-        xmin = cp.x;
-      }
-      if (cp.x > xmax) {
-        xmax = cp.x;
-      }
-      if (cp.z < zmin) {
-        zmin = cp.z;
-      }
-      if (cp.z > zmax) {
-        zmax = cp.z;
-      }
+    for (ChunkPosition cp : chunks) {
+      xmin = Math.min(cp.x, xmin);
+      xmax = Math.max(cp.x, xmax);
+      zmin = Math.min(cp.z, zmin);
+      zmax = Math.max(cp.z, zmax);
     }
-
     xmax += 1;
     zmax += 1;
     xmin *= 16;
     xmax *= 16;
     zmin *= 16;
     zmax *= 16;
 
-    int maxDimension = Math.max(yMax - yMin, Math.max(xmax - xmin, zmax - zmin));
-    int requiredDepth = QuickMath.log2(QuickMath.nextPow2(maxDimension));
+    return new BlockBounds(xmin, xmax, zmin, zmax);
+  }
+
+  static class BlockBounds {
+    final int xmin;
+    final int xmax;
+    final int zmin;
+    final int zmax;
 
-    if(centerOctree) {
-      int xroom = (1 << requiredDepth) - (xmax - xmin);
+    private BlockBounds(int xmin, int xmax, int zmin, int zmax) {
+      this.xmin = xmin;
+      this.xmax = xmax;
+      this.zmin = zmin;
+      this.zmax = zmax;
+    }
+  }
+
+  static Vector3i calculateOctreeOrigin(int yMax, int yMin, BlockBounds bounds,
+      boolean centerOctree) {
+    int requiredDepth = calculateOctreeRequiredDepth(bounds, yMax, yMin);
+
+    if (centerOctree) {
+      int xroom = (1 << requiredDepth) - (bounds.xmax - bounds.xmin);
       int yroom = (1 << requiredDepth) - (yMax - yMin);
-      int zroom = (1 << requiredDepth) - (zmax - zmin);
+      int zroom = (1 << requiredDepth) - (bounds.zmax - bounds.zmin);
 
-      origin.set(xmin - xroom / 2, -yroom / 2, zmin - zroom / 2);
+      return new Vector3i(bounds.xmin - xroom / 2, -yroom / 2, bounds.zmin - zroom / 2);
     } else {
       // Note: Math.floorDiv rather than integer division for round toward -infinity
-      origin.set(xmin, Math.floorDiv(yMin, 16) * 16, zmin);
+      return new Vector3i(bounds.xmin, Math.floorDiv(yMin, 16) * 16, bounds.zmin);
     }
-    return requiredDepth;
+  }
+
+  static int calculateOctreeRequiredDepth(BlockBounds bounds, int yMax, int yMin) {
```

Comment:
This could be moved into the `Octree` class

Reply:
Good point